<?xml version="1.0" ?>
<sdf version="1.6">
  <world name="battlebot_arena">

    <physics name="default_physics" default="true" type="ode">
  <max_step_size>0.001</max_step_size>
  <real_time_factor>1</real_time_factor>
  <real_time_update_rate>1000</real_time_update_rate>
  <ode>
    <solver>
      <type>quick</type>
      <iters>250</iters>
      <sor>1.3</sor>
      <use_dynamic_moi_rescaling>0</use_dynamic_moi_rescaling>
    </solver>
    <constraints>
      <cfm>0</cfm>
      <erp>0.2</erp>
      <contact_max_correcting_vel>100</contact_max_correcting_vel>
      <contact_surface_layer>0.005</contact_surface_layer>
    </constraints>
  </ode>
</physics>

    <light type="directional" name="sun">
      <cast_shadows>true</cast_shadows>
      <pose>0 0 10 0 0 0</pose>
      <diffuse>0.8 0.8 0.8 1</diffuse>
      <specular>0.2 0.2 0.2 1</specular>
      <direction>-0.5 0.1 -0.9</direction>
    </light>

    <model name="ground_plane">
      <static>true</static>
      <link name="link">
        <collision name="collision">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>10 10</size>
            </plane>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <plane>
              <normal>0 0 1</normal>
              <size>10 10</size>
            </plane>
          </geometry>
        </visual>
      </link>
    </model>

    <model name="arena_walls">
      <static>true</static>
      <link name="wall_link">

        <collision name="north_wall">
          <pose>0 5.5 0.5 0 0 0</pose>
          <geometry><box><size>11 1 1</size></box></geometry>
        </collision>
        <visual name="north_wall_visual">
          <pose>0 5.5 0.5 0 0 0</pose>
          <geometry><box><size>11 1 1</size></box></geometry>
          <material><diffuse>0.6 0.6 0.6 1</diffuse></material>
        </visual>

        <collision name="south_wall">
          <pose>0 -5.5 0.5 0 0 0</pose>
          <geometry><box><size>11 1 1</size></box></geometry>
        </collision>
        <visual name="south_wall_visual">
          <pose>0 -5.5 0.5 0 0 0</pose>
          <geometry><box><size>11 1 1</size></box></geometry>
          <material><diffuse>0.6 0.6 0.6 1</diffuse></material>
        </visual>

        <collision name="east_wall">
          <pose>5.5 0 0.5 0 0 0</pose>
          <geometry><box><size>1 11 1</size></box></geometry>
        </collision>
        <visual name="east_wall_visual">
          <pose>5.5 0 0.5 0 0 0</pose>
          <geometry><box><size>1 11 1</size></box></geometry>
          <material><diffuse>0.6 0.6 0.6 1</diffuse></material>
        </visual>

        <collision name="west_wall">
          <pose>-5.5 0 0.5 0 0 0</pose>
          <geometry><box><size>1 11 1</size></box></geometry>
        </collision>
        <visual name="west_wall_visual">
          <pose>-5.5 0 0.5 0 0 0</pose>
          <geometry><box><size>1 11 1</size></box></geometry>
          <material><diffuse>0.6 0.6 0.6 1</diffuse></material>
        </visual>

      </link>
    </model>

    <model name="moving_obstacle">
      <pose>0 0 0.25 0 0 0</pose>
      <static>false</static>
      <link name="link">
        <gravity>0</gravity>
        <inertial>
          <mass>200</mass>
          <inertia>
            <ixx>999</ixx>
            <iyy>999</iyy>
            <izz>999</izz>
            <ixy>0.0</ixy>
            <ixz>0.0</ixz>
            <iyz>0.0</iyz>
          </inertia>
        </inertial>
        <visual name="visual">
          <geometry><box><size>2 0.5 0.5</size></box></geometry>
          <material><diffuse>0.8 0.2 0.2 1</diffuse></material>
        </visual>
        <collision name="moving_obstacle_collision">
          <geometry><box><size>2 0.5 0.5</size></box></geometry>
          <surface>
            <friction>
              <ode>
                <mu>100.0</mu>
                <mu2>100.0</mu2>
              </ode>
            </friction>
            <contact>
            <ode>
              <kp>100000.0</kp> <kd>1.0</kd>
              <max_vel>0.1</max_vel>
              <min_depth>0.003</min_depth>
            </ode>
            </contact>
          </surface>
        </collision>

        <sensor name="contact_sensor" type="contact">
        <always_on>true</always_on>
        <update_rate>30</update_rate>
        <contact>
          <collision>moving_obstacle_collision</collision>
        </contact>
        <plugin name="contact_plugin" filename="libgazebo_ros_bumper.so">
          <ros>
            <namespace>/moving_obstacle</namespace>
            <remapping>bumper_states:=contact</remapping>
          </ros>
          <frame_name>moving_obstacle</frame_name>
        </plugin>
      </sensor>

      </link>
      
      <plugin name="planar_move_plugin" filename="libgazebo_ros_planar_move.so">
        <ros>
          <namespace>/moving_obstacle</namespace>
        </ros>
      </plugin>
    </model>

    <include>
      <uri>model://saw_blade</uri>
      <name>saw_blade_1</name>
      <pose>2.5 2.5 0.25 0 0 0</pose>
    </include>

    <include>
      <uri>model://saw_blade</uri>
      <name>saw_blade_2</name>
      <pose>-2.5 -2.5 0.25 0 0 0</pose>
    </include>

 
    <model name='healing_pack_1'>
      <static>true</static>
      <pose>4 4 0.15 0 0 0</pose>
      <link name='link'>
        <visual name='visual'>
          <geometry><box><size>0.5 0.5 0.3</size></box></geometry>
          <material><script><uri>file://media/materials/scripts/gazebo.material</uri><name>Gazebo/Blue</name></script></material>
        </visual>
        <collision name='collision'>
          <geometry><box><size>0.5 0.5 0.3</size></box></geometry>
        </collision>
        <sensor name='contact_sensor' type='contact'>
          <always_on>true</always_on>
          <update_rate>10</update_rate>
          <contact><collision>collision</collision></contact>
          <plugin name='contact_plugin' filename='libgazebo_ros_bumper.so'>
            <ros>
              <namespace>/healing_pack_1</namespace>
              <remapping>bumper_states:=contact</remapping>
            </ros>
            <frame_name>link</frame_name>
          </plugin>
        </sensor>
      </link>
    </model>

    <model name='healing_pack_2'>
      <static>true</static>
      <pose>-4 -4 0.15 0 0 0</pose>
      <link name='link'>
        <visual name='visual'>
          <geometry><box><size>0.5 0.5 0.3</size></box></geometry>
          <material><script><uri>file://media/materials/scripts/gazebo.material</uri><name>Gazebo/Blue</name></script></material>
        </visual>
        <collision name='collision'>
          <geometry><box><size>0.5 0.5 0.3</size></box></geometry>
        </collision>
        <sensor name='contact_sensor' type='contact'>
          <always_on>true</always_on>
          <update_rate>10</update_rate>
          <contact><collision>collision</collision></contact>
          <plugin name='contact_plugin' filename='libgazebo_ros_bumper.so'>
            <ros>
              <namespace>/healing_pack_2</namespace>
              <remapping>bumper_states:=contact</remapping>
            </ros>
            <frame_name>link</frame_name>
          </plugin>
        </sensor>
      </link>
    </model>

  </world>
</sdf>
